### ISSUE TITLE
Reproducing paper result

### ISSUE THREAD

**Issue by ivicts** [atomistic-machine-learning/schnetpack#77]
Hi,

Can you let me know what setting of parameters, optimizers, split and seed to reproduce the result of 
https://arxiv.org/pdf/1712.06113.pdf?

For homo, I tried:
python3 src/scripts/schnetpack_qm9_new.py train schnet data/seed_2000/homo/data/qm9.db data/seed_2000/homo/model --split 109000 1000 --cuda --property homo --seed 2000 --parallel

I got train/val/test mae as 0.00497/0.04701/0.04877, but the test mae stated in the paper is 0.041. What should I do to get this result? 

Should I change the split to "--split 110000 1000" and batch size to "--batch_size 32" following the arxiv paper?

**Comment by ktschuett** [comment]
Hi,
you should set "--aggregation_mode avg", since HOMO is an intensive property. Let me know if it doesn't work after that. The split and batch size should not matter that much.

@Stefaanhess We should add an example about this in the docs.

**Comment by ivicts** [comment]
Hi thanks for the reply, so I need to change  "--aggregation_mode avg" only for HOMO, LUMO and gap is it?

Is there any other configuration that I need to take note to reproduce the result of other properties in QM9?

I also get a better result running the code than the stated result on schnetpack paper for dipole moment. Is this expected?

It would be useful if you can add a table of result for every property of QM9 in the docs, so that people know if they reproduced and use the codes correctly. Seems like the result you have in arxiv paper are outdated and your schnetpack paper https://arxiv.org/abs/1809.01072?context=physics only give the result for U0 and dipole moment.

**Comment by ktschuett** [comment]
The results in the [software paper](https://pubs.acs.org/doi/full/10.1021/acs.jctc.8b00908)) should be closest. For the other properties, you should get errors in the same order of magnitude for all practical purposes.

The dipole should indeed be better, since we're using a layer based on latent charges as described in the software paper linked above.

**Comment by ivicts** [comment]
Hi,

I want to reproduce the result on all properties. There is some deviation of the result for some of the properties:
The unit is in ev.

| Properties | Schnet paper | Reproduce Result |
|------------|--------------|------------------|
| gap        | 0.063        | 0.06708          |
| HOMO       | 0.041        | 0.04483          |
| LUMO       | 0.034        | 0.03666          |
| R2         | 0.073        | 0.21775          |
| U0         | 0.011        | 0.01442          |

I have used "--aggregation_mode avg" for HOMO, LUMO, and Gap. Do you think the small deviation of Gap, HOMO, LUMO and U0 when I reproduce the result makes sense and I reproduced it correctly?

What about R2? What should I change to reproduced the paper?

For other properties, the result is more or less the same as the paper.

**Comment by ktschuett** [comment]
Hi,
regarding R2, this is probably due to unit conversion. We use the standard ASE units, which is eV and Angstrom. The result in the paper is in the original QM9 unit Bohr^2, i.e. 0.21775 Ang^2 -> 0.061 Bohr^2

The other results seem relatively close and might be due to the split. I will check what I get.

**Comment by ivicts** [comment]
Hi,

I forgot the unit conversion.
How did you do the conversion? I got this from Google:
![image](https://user-images.githubusercontent.com/19903139/55221251-006e6780-5244-11e9-9bd3-eea277bec09e.png)

I used seed = 2000 if you want to reproduce my experiments.

Alpha also seems to be much better than the one stated in the paper, I have converted the unit to Bohr^3:

| Properties | Schnet paper | Reproduce Result |
|------------|--------------|------------------|
| alpha      | 0.235        | 0.09299204951    |
| R2         | 0.073        | 0.21775          |
| mu         | 0.02         | 0.02516879249    |

Maybe there is a problem in the unit conversion?

I got:
Angstrom^2 -> Bohr^2: Multiply by 3.571079473
Angstrom^3 -> Bohr^3: Multiply by 6.748376012
For Mu:
Angstrom E -> Debye: Multiply by 4.803204673

**Comment by ktschuett** [comment]
Oh, sry, you're right, did it the wrong way around. Would have fitted too well.

The dipole moment seems to fit up to the small discrepancy, which will likely have a similar reason as the other properties. I'm guessing learning rate schedule. Experiments are running.

Alpha and R2 might indeed be some unit issue. I would say that the polarizability result is correct, since I remember that I had some trouble predicting it with the old code. I will look into this.

**Comment by ktschuett** [comment]
Hi,
so I did some digging in my old results. I am confident now that they are correct and that there is no conversion error. I still had the predicted values stored. The errors in the different splits are:

|Property | split 1 | split 2 | split3 | avg |
|-----------|-----------|---------|---------|------|
| alpha | 0.125 | 0.246 | 0.335 | 0.235 |
| r^2 | 0.076 | 0.073 | 0.069  | 0.073 |

There were, as I mentioned, sometimes problems during training with the polarizability. Maybe due to the small batches. It would be great if this works now reliably in SchNetPack. 

I am still not sure what the problem with the spatial extent is. With SchNetPack, I get similar values as you get. In the paper, I used 64 features and no cutoff. Also the weight initialization was slightly different. At some point the weights of the last layers of the interaction residual were initialized to zero. This could also slightly affect the other results since I changed this only recently. 

My current theory is that r^2 needs a larger cutoff. It would make sense given what the property describes. I will start some experiment.

**Comment by ivicts** [comment]
Hi, thanks for the update! I will wait until you resolve the issue with alpha and R2.

btw, have you you checked the mu, homo, lumo, gap and U0? May I confirm I reproduce it correctly?
> The other results seem relatively close and might be due to the split. I will check what I get.

**Comment by ktschuett** [comment]
We just found a bug in the HardCutoff funtion (PR #86) that got introduced after the software paper. That could be very well the reason for the difference in performance. I'll keep you posted.

**Comment by ktschuett** [comment]
While the models are still running (now with fixed cutoff), I think the results should be telling. I also added the possibility of the choosing a cosine cutoff function to the scripts, since this is what we are now. Here are the validation errors:

| Property | cutoff | cutoff_function (PR #91) | Validation error |
|------------|----------|---------------------------------|-----------------------|
| U0 | 5 Å | hard | 0.0114 eV |
| U0 | 5 Å | cosine | 0.0078 eV |
| R^2 | 20 Å | hard | 0.0084 Bohr^2 |

Here are argument files:
[U0_hard](https://github.com/atomistic-machine-learning/schnetpack/files/3037595/U0_args.txt)
[U0_cutoff.txt](https://github.com/atomistic-machine-learning/schnetpack/files/3037598/U0_cutoff.txt)
[R2_args.txt](https://github.com/atomistic-machine-learning/schnetpack/files/3037601/R2_args.txt)

I.e., now running schnetpack should reproduce the paper results close enough. Of course changing batch size, learning rate schedule makes some difference, but especially since we get much better results with the cosine cutoff (at least for U0, we need to check with the other properties), I wouldn't take it too seriously. 


I created an issue to pick up on your suggestion of putting benchmark results in the docs. Thank you for this! We are working towards the next stable release at the moment, so we will create them whenever we're releasing. Then we will notice in future versions whether we introduced some weird bug again ;)

**Comment by L3viAckerman** [comment]
Hi @ktschuett , I am trying to reproduce the paper results too, but I'm facing with that time need for training model is too much, about 130s for one epoch. So when I read the params, I see epochs is 5000, so huge for my device (I only have 1 K40m GPU). I wonder what number of epochs is the min  for training model to reproduce the paper results. Thanks you so much!
